src/module_4/model_api.py

```diff
@@ -0,0 +1,215 @@
+import json
+import pandas as pd
+import os
+import pickle
+from sklearn.pipeline import Pipeline
+from typing import List, NamedTuple
+
+
+class ExpectedEventKeysFit(NamedTuple):
+    model_parametrisation: dict
+    train_cols: list[str]
+    target: str
+    data: any
+    pipe: str
+    pipeline: callable
+    date: str
+
+
+class ExpectedEventKeysPredict(NamedTuple):
+    date: str
+    users: str
+
+
+def perim_modelo(df: pd.DataFrame, min_items: int = 5) -> pd.DataFrame:
+    """
+    Given a df that has columns id | order_id it generates a new dataframe which
+    contains the orders that have more items than the argument min_items
+    """
+    assert {"variant_id", "order_id"}.issubset(set(df.columns))
+    order_counts = df.groupby("order_id").size().reset_index(name="item_count")
+    order_counts_filtered = order_counts[order_counts["item_count"] >= min_items]
+    df = df.merge(order_counts_filtered, on="order_id", how="inner")
+    return df
+
+
+def col_to_datetime(df: pd.DataFrame, col_name: str) -> pd.DataFrame:
+    assert col_name in df.columns
+    try:
+        df[col_name] = pd.to_datetime(df[col_name])
+    except:
+        print(f"Could not change types at {df.head()}, with column {col_name}")
+    return df
+
+
+def data_pipe(df: pd.DataFrame) -> pd.DataFrame:
+    df = (
+        df.pipe(perim_modelo, 5)
+        .pipe(col_to_datetime, "order_date")
+        .pipe(col_to_datetime, "created_at")
+    )
+    return df
+
+
+def serialize_model(model: Pipeline, name: str, prefix_path=None) -> None:
+    if prefix_path is not None:
```

Comment:
You can achieve almost the same behavior with ```if prefix_path:```. The only difference is that the code I showed you also returns False when empty string.